src/storage/src/hummock/store/version.rs

```diff
@@ -133,13 +152,8 @@ pub enum VersionUpdate {
 
 #[derive(Clone)]
 pub struct StagingVersion {
-    // newer data comes first
-    // Note: Currently, building imm and writing to staging version is not atomic, and therefore
-    // imm of smaller batch id may be added later than one with greater batch id
-    pub imm: VecDeque<ImmutableMemtable>,
-
-    // newer data comes first
-    pub sst: VecDeque<Arc<StagingSstableInfo>>,
+    // newer data comes in last, so we need to reverse them in read.
```

Comment:
Instead of `Vec`, I think we can store as `BTreeMap<ImmId, StagingData>` so that we can replace the imm with sst with less cost while maintain the order of data.

As for the `ImmId` key of `StagingData`, for imm, the key is the imm id of the imm itself. For sst, it can be the greatest (or least) imm id of the imm of this shard. For example, let's say a sst contains the data of imm of id [1, 2] from instance A, and imm of id [3, 4] from instance B. Then in the read version of A, its key is 2, and in the read version of B, its key is 4. When we replace the imms with the sst, we only need to update the read version A and B. When updating read version A, we can simply remove 1 and 2, and insert the sst with key 1. In this way, we can even add assertion that the upload task does not skip some imm. (For example, we can assert that there is no imm between 1 and 2 when we do the replacement)

In each `StagingSstableInfo`, we currently maintain `Vec<ImmId>` to represent that it holds the data of these imm id. We can change to maintain `HashMap<LocalInstanceId, Vec<ImmId>>`.

Reply:
Emmm ? I always thought the number of `imm` would not be too many....